rct_optimizations/test/include/rct_optimizations_tests/pose_generator.h

```diff
@@ -25,70 +26,258 @@ struct PoseGenerator
+                                 unsigned phi_cnt_,
+                                 double z_rot_,
+                                 const Eigen::Isometry3d& target_offset_)
+    : HemispherePoseGenerator(r_, theta_cnt_, phi_cnt_, z_rot_, z_rot_, target_offset_)
   {
+
+  }
+
+  inline HemispherePoseGenerator(double r_, unsigned theta_cnt_, unsigned phi_cnt_)
+    : HemispherePoseGenerator(r_, theta_cnt_, phi_cnt_, 0.0, Eigen::Isometry3d::Identity())
+  {
+
+  }
+
+  inline HemispherePoseGenerator()
+    : HemispherePoseGenerator(2.0, 10, 10, 0.0, Eigen::Isometry3d::Identity())
+  {
+
   }
 
   virtual std::vector<Eigen::Isometry3d> generate(
-    const Eigen::Vector3d &target_origin) const override final;
+    const Eigen::Isometry3d &target_origin) override final;
+
+  virtual double getZRot() override final;
+
+  double r; /** @brief Radius of the hemisphere */
+  unsigned theta_cnt; /** @brief The number of points in the theta-wise direction*/
+  unsigned phi_cnt; /** @brief The number of points in the phi-wise direction */
+  Eigen::Isometry3d target_offset; /** @brief Transform from target origin to origin of pose pattern */
 
-  double r; /* @brief Radius of the hemisphere */
-  unsigned theta_cnt; /* @brief The number of points in the theta-wise direction*/
-  unsigned phi_cnt; /* @brief The number of points in the phi-wise direction */
+private:
+  std::mt19937 mt_rand; /** @brief Psuedo-random number generator */
+  std::uniform_real_distribution<double> z_sampler; /** @brief Sampler for camera Z+ rotation */
+};
+
+/**
+ * @brief Generates camera poses in a hemisphere pattern. Randomly vary rotation around the camera Z+ axis.
+ */
+struct RandomZRotHemispherePoseGenerator : HemispherePoseGenerator
+{
+  inline RandomZRotHemispherePoseGenerator(double r_,
+                                           unsigned theta_cnt_,
+                                           unsigned phi_cnt_,
+                                           const Eigen::Isometry3d& target_offset_)
+    : HemispherePoseGenerator (r_, theta_cnt_, phi_cnt_, 0.0, 2 * M_PI, target_offset_)
+  {
+
+  }
+
+  inline RandomZRotHemispherePoseGenerator()
+    : RandomZRotHemispherePoseGenerator(2.0, 10, 10, Eigen::Isometry3d::Identity())
+  {
+
+  }
 };
 
 /**
  * @brief Generates camera positions using a conical template, with the target at the 'point'
  */
 struct ConicalPoseGenerator : PoseGenerator
 {
-  inline ConicalPoseGenerator(double r_ = 1.0,
-                              double h_ = 2.0,
-                              unsigned n_poses_ = 20)
+protected:
+  inline ConicalPoseGenerator(double r_,
+                              double h_,
+                              unsigned n_poses_,
+                              double z_rot_min_,
+                              double z_rot_max_,
+                              const Eigen::Isometry3d& target_offset_,
+                              unsigned long seed_ = std::mt19937::default_seed)
```

Comment:
Same comments about moving the random elements to the specific random implementation